File: python/베스트앨범.py
from typing import List
from collections import defaultdict


def solution(genres: List[str], plays: List[int]) -> List[int]:
    # 장르 별로 총 재생된 횟수를 보관
    top_genres = defaultdict(int)
    # 각각의 음악 정보(고유번호, 재생된 횟수)를 카테고리(장르)별로 보관
    category = defaultdict(list)

    for id in range(len(genres)):
        genre, play_num = genres[id], plays[id]
        top_genres[genre] += play_num
        category[genre].append(
            (id, play_num)
        )

    print(f'top_genres: {dict(top_genres)}')
    print(f'category: {dict(category)}')

    # 가장 많이 재생된 장르가 앞에 오도록 내림차순
    sorted_top_genres = sorted(
        top_genres,
        key=lambda x: top_genres[x],
        reverse=True
    )
    print(f'sorted_top_genres: {sorted_top_genres}')

    answer = []
    for genre in sorted_top_genres:
        # 재생 횟수로 내림차순 정렬, 재생 횟수가 같으면 고유번호 기준으로 오름차순 정렬
        sorted_genre = sorted(
            category[genre],
            key=lambda x: (-x[-1], x[0])
        )
        print(f'genre: {genre}, sorted_genre: {sorted_genre}')
        # 상위 2개 이하의 노래만 순서대로 id를 추출하여 리스트로 만듦
        answer += [
            id for (id, count) in sorted_genre[:2]
        ]
        print(f'answer: {answer}')

    return answer


# def solution(genres: List, plays: List) -> List:
#     genres_plays = []
#
#     for item in zip(genres, plays):
#         genres_plays.append(item)
#
#     genres_plays.sort(key=lambda music: (music[0], -music[-1]))
#
#     print(genres_plays)
#
#     answer = []
#     return answer


if __name__ == "__main__":
    genres = ["classic", "pop", "classic", "classic", "pop"]
    plays = [500, 600, 150, 800, 2500]
    print(solution(genres, plays))

File: Joowon/JavaScript/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    let map = new Map();
    let mapForSize = new Map();

    // 전체 재생횟수 카운팅
    genres.forEach((category, i) => {
        const num = plays[i];
        map.get(category) ? map.set(category, map.get(category) + num) : map.set(category, num);
    });

    answer = genres.map((category, i) => {
        return {
            category: category,
            num: plays[i],
            index: i
        };
    }).sort((a, b) => {
        if (a.category !== b.category) return map.get(b.category) - map.get(a.category);
        if (a.num !== b.num) return b.num - a.num;
        return a.index - b.index;
    }).filter(ele => {
        if (mapForSize.get(ele.category)) {
            if (mapForSize.get(ele.category) < 2) {
                mapForSize.set(ele.category, 2);

                return true;
            } else {
                return false;
            }
        } else {
            mapForSize.set(ele.category, 1);

            return true;
        }
    }).map(ele => ele.index);
    

    return answer;
}

console.log(solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500])); // 답: [4, 1, 3, 0]
console.log(solution(["classic", "pop", "classic", "classic", "pop", "test"], [500, 600, 150, 800, 2500, 100])); // 답: [4, 1, 3, 0, 5]
console.log(solution(["classic", "pop", "classic", "classic", "pop", "test"], [500, 600, 150, 800, 600, 100])); // 답: [3, 0, 1, 4, 5]
console.log(solution(["classic", "classic", "classic", "classic", "classic"], [5, 5, 5, 5, 5])); // 답: [0, 1]
console.log(solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 500, 800, 2500])); // 답: [4, 1, 3, 0]
console.log(solution(["pop", "pop", "pop", "rap", "rap"], [45, 50, 40, 60, 70])); // 답: [1, 0, 4, 3]
console.log(solution(["classic"], [300])); // 답: [0]